keras_rs/src/utils/ranking_metrics_utils.py

```diff
@@ -0,0 +1,240 @@
+      nonzero relevance and nonzero weights is proportional to
+
+      ```
+      per_list_weights / sum(per_list_weights) *
+      num(sum(relevance) != 0) / num(lists)
+      ```
+
+      The rest have weights 1.0 / num(lists).
+
+    Args:
+        weights:  The weights `Tensor` of shape [batch_size, list_size].
+        relevance:  The relevance `Tensor` of shape [batch_size, list_size].
+
+    Returns:
+        A tensor of shape [batch_size, 1], containing the per-list weights.
+    """
+    # TODO: Pretty nasty function. Check if we can simplify it at a later point.
+
+    # Calculate if the sum of weights per list is greater than 0.0.
+    nonzero_weights = ops.greater(ops.sum(weights, axis=1, keepdims=True), 0.0)
+    # Calculate the sum of relevance per list
+    per_list_relevance = ops.sum(relevance, axis=1, keepdims=True)
+
+    # Identify lists where both weights and relevance sums are non-zero.
+    nonzero_relevance_condition = ops.greater(per_list_relevance, 0.0)
+    nonzero_relevance = ops.where(
+        nonzero_weights,
+        ops.cast(nonzero_relevance_condition, "float32"),
+        ops.zeros_like(per_list_relevance),
+    )
+    # Count the number of lists with non-zero relevance (and implicitly
+    # non-zero weights).
+    nonzero_relevance_count = ops.sum(nonzero_relevance, axis=0, keepdims=True)
+
+    # Calculate the per-list weights using the core formula
+    # Numerator: sum(weights * relevance) per list
+    numerator = ops.sum(weights * relevance, axis=1, keepdims=True)
+    # Denominator: per_list_relevance = sum(relevance) per list
+    per_list_weights = ops.divide_no_nan(numerator, per_list_relevance)
+
+    # Calculate the sum of the initially computed per-list weights (where
+    # relevance > 0)
+    sum_weights = ops.sum(per_list_weights, axis=0, keepdims=True)
+
+    # Calculate the average weight to use as default for lists with zero
+    # relevance but non-zero weights. If no lists have non-zero relevance,
+    # default to 1.0.
+    avg_weight = ops.where(
+        ops.greater(nonzero_relevance_count, 0.0),
+        ops.divide_no_nan(sum_weights, nonzero_relevance_count),
+        ops.ones_like(nonzero_relevance_count),
+    )
+
+    # Final assignment of weights based on conditions:
+    # 1. If sum(weights) == 0 for a list, the final weight is 0.
+    # 2. If sum(weights) > 0 AND sum(relevance) > 0, use the calculated
+    # `per_list_weights`.
+    # 3. If `sum(weights) > 0` AND `sum(relevance) == 0`, use the calculated
+    # `avg_weight`.
+    final_weights = ops.where(
+        nonzero_weights,
+        ops.where(
+            nonzero_relevance_condition,
+            per_list_weights,
+            ops.ones_like(per_list_weights) * avg_weight,
+        ),
+        ops.zeros_like(per_list_weights),
+    )
+
+    return final_weights
+
+
+def default_gain_fn(label: types.Tensor) -> types.Tensor:
+    return ops.subtract(ops.power(2.0, label), 1.0)
```

Comment:
nitpick: I have always assumed `ops.square` is faster.

Reply:
Will address these in a new PR for DCG, nDCG.